### PR TITLE
When node is clicked, it's brought at the top position.

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
[MAGN-9212](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9212) When a node is in focus, its UI and the preview bubble should be 'brought to the front' of the view order so that it is easy to review and pin

### Declarations

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.

### Reviewers

@ramramps

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs [hunk 1]
```diff
@@ -428,33 +440,33 @@ private void OnPreviewControlStateChanged(object sender, EventArgs e)
             switch (preview.CurrentState)
             {
                 case PreviewControl.State.Hidden:
-                {
```
Comment (aosyatnik) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs]: It's just indents after Ctrl+E+F

**Comment by ramramps** [comment]
@aosyatnik  - this is looking very good. Few things 
1. Clicking on the preview bubble should bring it to the top. This includes the items inside the bubble. 
2. Fix the animation when you click on the preview bubble (when you change the z-order)
3. If a node is under the preview bubble, you should be able to click anywhere on the node to bring that node to the top (that is reordering the z-index). today, the bubble has a border (or something similiar) preventing that.

**Comment by aosyatnik** [comment]
Hi, @ramramps , I fixed some problems. Here is my update:
1. I used preview event to bring in front node even if we click on preview bubble. This time I checked connectors several times :)
2. This will be fixed as part of https://github.com/DynamoDS/Dynamo/pull/5951 . The reason of preview animation is that Canvas captures mouse.
3. Yes, there is border, but not around preview bubble. There is invisible border around each port. This border prevents clicking on node. Here it is:
image image

This transparent rectangle is used for snapping ports. So, when you click on it, it handles mouse click. In order to "click through ui elements" we need to set `IsHitTestVisible` to false. (Ports.xaml row 82). But in this case we will lose port snapping. So what I did... I bind `IsHitTestVisible` to PortViewModel. `IsHitTestVisible` is updated, when `ActiveConnector` is changed. That means when there is no active connector, we can click though invisible rectangle. But as soon as active connector is not null, invisible rectangle becomes snapping area.

**Comment by jnealb** [comment]
@aosyatnik @ramramps This may be relevant to your solution: "MAGN-8085  After a node selection some right clicks in the canvas are discarded".

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs [hunk 2]
```diff
@@ -71,6 +71,11 @@ public bool UsingDefaultValue
             set { _port.UsingDefaultValue = value; }
         }
 
+        public bool IsHitTestVisible
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs]: Comments.

**Comment by ramramps** [comment]
@aosyatnik -  We will discuss this after the stand up. I would like to show the behavior.

**Comment by aosyatnik** [comment]
Hi @ramramps , I believe I resolved all found issues. Let's take a look.

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs [hunk 3]
```diff
@@ -23,12 +23,22 @@ namespace Dynamo.Controls
...
         private PreviewControl previewControl = null;
         private const int previewDelay = 1000;
 
         /// <summary>
+        /// ZIndex is used to order nodes, when some node is clicked.
+        /// This selected node should be moved above others.
+        /// Start value of zIndex is 3, because 1 is for groups and 2 is for connectors.
+        /// Nodes should be always at the top.
+        /// </summary>
+        private static uint zIndex = 3;
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs]: Why can't this be int? Int32.Maxvalue is sufficient here.  uint is not cls complaint and I don't see any specific use of uint.

**Comment by ramramps** [comment]
@aosyatnik @Racel -  this scennario: pin the previewbubble. click the node below the preview bubble, view order of the node is changed,  but not the connector. It looks like the node does not have connector.

image

**Comment by aosyatnik** [comment]
@ramramps , fixed it. Look `ConnectorViewModel`. I set the same ZIndex for ConnectorViewModel as for NodeViewModel. However, there wasn't acceptance criteria, that connector should be brought in front... But I believe now it's better.

File: src/DynamoCoreWpf/ViewModels/Core/ConnectorViewModel.cs [hunk 4]
```diff
@@ -188,7 +194,7 @@ public bool PlineVisibility
             }
         }
 
-        public NodeViewModel Nodevm
+        public NodeViewModel StartNode
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Core/ConnectorViewModel.cs]: Comments?  It's hard to understand what is a start node and end node. So, please write a brief comment.

File: src/DynamoCoreWpf/ViewModels/Core/ConnectorViewModel.cs [hunk 5]
```diff
@@ -59,10 +60,15 @@ public double Top
             get { return 0; }
         }
 
-        //Changed the connectors ZIndex to 2. Groups have ZIndex of 1.
-        public double ZIndex
+        private int zIndex = Configurations.StartZIndexForConnector;
+        public int ZIndex
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Core/ConnectorViewModel.cs]: comment.

**Comment by ramramps** [comment]
@aosyatnik - Another one if you have more than one port
image

**Comment by Racel** [comment]
@ramramps @aosyatnik - @mccrone and I reviewed the two options above (wires below preview bubble and wires above). We came to the same conclusion that @aosyatnik and I came to this morning, which is that the old behavior makes more sense with the wires being below the preview bubble. @mccrone pointed out that this is the standard normal interaction in other visual scripting programs. So, I think we should maintain the existing paradigm.

Additionally, the conflicts that arose in the second image are not desirable, and it will likely be extremely difficult to determine which node each wire will adopt its z-index from. Will it be from the input node or the output node? 

In summary, @ramramps and @aosyatnik, I suggest that we revert back to the existing behavior where wires are also shown below the preview bubble regardless of the node's view order. Wires should be treated as "second-class" citizens.

**Comment by aosyatnik** [comment]
@ramramps , @Racel  I reverted changes of ConnectorViewModel. If you think it's ok, then set LGTM, but don't merge. I'll create new PR with 1 commit.

**Comment by ramramps** [comment]
@aosyatnik - LGTM.

**Comment by aosyatnik** [comment]
Replaced by https://github.com/DynamoDS/Dynamo/pull/5990
